cocos/core/math/mat4.ts

```diff
@@ -828,30 +828,44 @@ export class Mat4 extends ValueType {
      * @en Extracts the scale, rotation and translation from the matrix, assuming it's composed in order of scale, rotation, translation
      * @zh 提取旋转、位移、缩放信息， 默认矩阵中的变换以 S->R->T 的顺序应用
      */
-    public static toRTS <InType extends IMat4Like, VecLike extends IVec3Like> (m: InType, q: Quat, v: VecLike, s: VecLike) {
-        s.x = Vec3.set(v3_1, m.m00, m.m01, m.m02).length();
-        m3_1.m00 = m.m00 / s.x;
-        m3_1.m01 = m.m01 / s.x;
-        m3_1.m02 = m.m02 / s.x;
-        s.y = Vec3.set(v3_1, m.m04, m.m05, m.m06).length();
-        m3_1.m03 = m.m04 / s.y;
-        m3_1.m04 = m.m05 / s.y;
-        m3_1.m05 = m.m06 / s.y;
-        s.z = Vec3.set(v3_1, m.m08, m.m09, m.m10).length();
-        m3_1.m06 = m.m08 / s.z;
-        m3_1.m07 = m.m09 / s.z;
-        m3_1.m08 = m.m10 / s.z;
-        const det = Mat3.determinant(m3_1);
-        if (det < 0) { s.x *= -1; m3_1.m00 *= -1; m3_1.m01 *= -1; m3_1.m02 *= -1; }
-        Quat.fromMat3(q, m3_1); // already normalized
-        Vec3.set(v, m.m12, m.m13, m.m14);
+    public static toRTS<InType extends IMat4Like, VecLike extends IVec3Like> (m: InType, q: Quat | null, v: VecLike | null, s: VecLike | null) {
+        const sx = Vec3.set(v3_1, m.m00, m.m01, m.m02).length();
+        const sy = Vec3.set(v3_1, m.m04, m.m05, m.m06).length();
+        const sz = Vec3.set(v3_1, m.m08, m.m09, m.m10).length();
+        if (s) {
+            s.x = sx;
+            s.y = sy;
+            s.z = sz;
+        }
+        if (v) {
+            Vec3.set(v, m.m12, m.m13, m.m14);
+        }
+        if (q) {
+            m3_1.m00 = m.m00 / sx;
+            m3_1.m01 = m.m01 / sx;
+            m3_1.m02 = m.m02 / sx;
+            m3_1.m03 = m.m04 / sy;
+            m3_1.m04 = m.m05 / sy;
+            m3_1.m05 = m.m06 / sy;
+            m3_1.m06 = m.m08 / sz;
+            m3_1.m07 = m.m09 / sz;
+            m3_1.m08 = m.m10 / sz;
+            const det = Mat3.determinant(m3_1);
+            if (det < 0) {
+                if (s) s.x *= -1;
```

Comment:
Whether getting q or note affects the value of s.

Reply:
fixed